propulsion-hotel/Reactor/SourceArgs.fs

```diff
@@ -0,0 +1,119 @@
+        | [<AltCommandLine "-cp">]          CheckpointConnectionString of string
+        | [<AltCommandLine "-s">]           Schema of string
+        | [<AltCommandLine "-b"; Unique>]   BatchSize of int
+        | [<AltCommandLine "-Z"; Unique>]   FromTail
+        interface IArgParserTemplate with
+            member a.Usage = a |> function
+                | ConnectionString _ ->     $"Connection string for the postgres database housing message-db when writing. (Optional if environment variable {CONNECTION_STRING} is defined)"
+                | ReadConnectionString _ -> $"Connection string for the postgres database housing message-db when reading. (Defaults to the (write) Connection String; Optional if environment variable {READ_CONN_STRING} is defined)"
+                | CheckpointConnectionString _ -> "Connection string used for the checkpoint store. If not specified, defaults to the connection string argument"
+                | Schema _ ->               $"Schema that should contain the checkpoints table Optional if environment variable {SCHEMA} is defined"
+                | BatchSize _ ->            "maximum events to load in a batch. Default: 100"
+                | FromTail _ ->             "(iff the Consumer Name is fresh) - force skip to present Position. Default: Never skip an event."
+
+    type Arguments(c : Args.Configuration, p : ParseResults<Parameters>) =
+        let writeConnStr =                  p.TryGetResult ConnectionString |> Option.defaultWith (fun () -> c.MdbConnectionString)
+        let readConnStr =                   p.TryGetResult ReadConnectionString |> Option.orElseWith (fun () -> c.MdbReadConnectionString) |> Option.defaultValue writeConnStr
+        let checkpointConnStr =             p.TryGetResult CheckpointConnectionString |> Option.defaultValue writeConnStr
+        let schema =                        p.TryGetResult Schema |> Option.defaultWith (fun () -> c.MdbSchema)
+        let fromTail =                      p.Contains FromTail
+        let batchSize =                     p.GetResult(BatchSize, 100)
+        let tailSleepInterval =             TimeSpan.FromMilliseconds 500.
+        member _.Connect() =
+                                            let connStrWithoutPassword = Npgsql.NpgsqlConnectionStringBuilder(checkpointConnStr, Password = null)
+                                            Log.Information("MessageDb connection {connectionString}", connStrWithoutPassword.ToString())
+                                            let client = Equinox.MessageDb.MessageDbClient(writeConnStr, readConnStr)
+                                            Equinox.MessageDb.MessageDbContext(client, batchSize)
+        member _.MonitoringParams(log : ILogger) =
+            log.Information("MessageDbSource batchSize {batchSize} Checkpoints schema {schema}", batchSize, schema)
+            if fromTail then log.Warning("(If new projector group) Skipping projection of all existing events.")
```

Comment:
Can we log this only if relevant somehow? i.e. if there's no checkpoint stored and fromTail is true

Reply:
This is a once-off override argument for when you make a new projection and are sure you don't want to replay, so it is not as noisy/painful as one might imagine

in Cosmos CFP, there is no API way to influence or determine a checkpoint has happened

`propulsion checkpoint` also provides ways to force checkpoints and/or set them to explicit values

I have not thought about the problem across all stores enough recently to have an immediate yes/no on whether that would be possible and/or desirable

(Propulsion.EventStore had a predecessor design which works more like you'd imagine - if this was to be resolved, the solution should have analogous behaviors for Propu;sion.EventStoreDb too)